src/control/server/storage/bdev.go

```diff
@@ -82,41 +144,82 @@ func (nch *NvmeHealth) TempF() float32 {
 	return (nch.TempC() * (9.0 / 5.0)) + 32.0
 }
 
-type (
-	// NvmeNamespace represents an individual NVMe namespace on a device and
-	// mirrors C.struct_ns_t.
-	NvmeNamespace struct {
-		ID   uint32 `json:"id"`
-		Size uint64 `json:"size"`
-	}
-
-	// SmdDevice contains DAOS storage device information, including
-	// health details if requested.
-	SmdDevice struct {
-		UUID       string      `json:"uuid"`
-		TargetIDs  []int32     `hash:"set" json:"tgt_ids"`
-		State      string      `json:"state"`
-		Rank       system.Rank `json:"rank"`
-		TotalBytes uint64      `json:"total_bytes"`
-		AvailBytes uint64      `json:"avail_bytes"`
-		Health     *NvmeHealth `json:"health"`
-		TrAddr     string      `json:"tr_addr"`
-	}
-
-	// NvmeController represents a NVMe device controller which includes health
-	// and namespace information and mirrors C.struct_ns_t.
-	NvmeController struct {
-		Info        string           `json:"info"`
-		Model       string           `json:"model"`
-		Serial      string           `hash:"ignore" json:"serial"`
-		PciAddr     string           `json:"pci_addr"`
-		FwRev       string           `json:"fw_rev"`
-		SocketID    int32            `json:"socket_id"`
-		HealthStats *NvmeHealth      `json:"health_stats"`
-		Namespaces  []*NvmeNamespace `hash:"set" json:"namespaces"`
-		SmdDevices  []*SmdDevice     `hash:"set" json:"smd_devices"`
+// NvmeNamespace represents an individual NVMe namespace on a device and
+// mirrors C.struct_ns_t.
+type NvmeNamespace struct {
+	ID   uint32 `json:"id"`
+	Size uint64 `json:"size"`
+}
+
+// SmdDevice contains DAOS storage device information, including
+// health details if requested.
+type SmdDevice struct {
+	UUID       string      `json:"uuid"`
+	TargetIDs  []int32     `hash:"set" json:"tgt_ids"`
+	State      BioState    `json:"-"`
+	Rank       system.Rank `json:"rank"`
+	TotalBytes uint64      `json:"total_bytes"`
+	AvailBytes uint64      `json:"avail_bytes"`
+	Health     *NvmeHealth `json:"health"`
+	TrAddr     string      `json:"tr_addr"`
+}
+
+// MarshalJSON marshals SmdDevice to JSON.
+func (sd *SmdDevice) MarshalJSON() ([]byte, error) {
```

Comment:
I don't think you need the custom marshal/unmarshal here -- you should be able to put them on the state type(s).

Reply:
Can you explain the disadvantage of having the Marshal/Unmarshal method on SmdDevice please?

Reply:
strangely seems to be working fine without any Marshal/UnMarshal methods, will leave without

Reply:
I think the json package uses reflect under the hood to figure out that the type is just a uint32 so it's able to figure it out. Usually you only need the custom marshal/unmarshal when you're translating between fundamental types (e.g. int/string).